probe.c

```diff
@@ -61,120 +69,161 @@ void probe_init()
-  st_reset();
-
-  // Plan moition. Planner buffer should be empty, bypass mc_line() 
-  plan_buffer_line(target, probe_rate, false, LINENUMBER_EMPTY_BLOCK); 
-
-  // Tell the system we are probing
-  probe.isprobing = 1;
-
-  // Enable hard limit checking
-  limits_enable(LIMIT_MASK & HARDSTOP_MASK, 0);
 
+    // Report the position where the probe was found
+    //TODO report_probe_parameters() Maybe rather in probe_loop or probe_move_to_sensor
+  }
 
+  // Check for ESTOP
+  if (ESTOP_PIN & ESTOP_MASK) {
+    sys.alarm |= ALARM_ESTOP;
+    SYS_EXEC |= (EXEC_FEED_HOLD | EXEC_ALARM | EXEC_CRIT_EVENT);
+  }
 }
 
-void probe_loop()
+uint8_t probe_loop()
 {
   // Start stepper
   st_prep_buffer();
   st_wake_up();
 
+  SYS_EXEC |= EXEC_CYCLE_START;
+
+  // Stay in this loop until alarm or until
+  // the probe is found. probe_check() is called
+  // from the stepper ISR to check if the active
+  // probe is found.
   while (probe.isprobing) {
     // Check for user reset and allow
     // protocol_execute_runtime to run in this loop 
     protocol_execute_runtime();
     
+    if(sys.abort)
+      return false;
+
     if (SYS_EXEC & EXEC_RESET) {
       protocol_execute_runtime();
-      return;
+      return false;
     } 
 
     if (ESTOP_PIN & ESTOP_MASK) {
       sys.alarm |= ALARM_ESTOP;
       SYS_EXEC |= (EXEC_FEED_HOLD | EXEC_ALARM | EXEC_CRIT_EVENT);
       protocol_execute_runtime();
-      return;
+      return false;
     }
 
     // Check if we never reach probe.
     if ((SYS_EXEC & EXEC_CYCLE_STOP) ) {
       sys.alarm |= ALARM_PROBE_FAIL;
       SYS_EXEC |= EXEC_CRIT_EVENT;
       protocol_execute_runtime();
-      return;
+      return false;
     }
   }
 
-  // Disable limits
-  limits_disable();
+  return true;
+
+}
+
+void probe_move_to_sensor(float * target, float feed_rate, uint8_t invert_feed_rate,
+  linenumber_t line_number, enum e_sensor sensor)
+{
+  // Set the active probe
+  probe.active_sensor = sensor;
+
+  if (sys.state != STATE_CYCLE)
+    protocol_auto_cycle_start();
+  
+  // Finish all queued commands
+  protocol_buffer_synchronize();
+
+  // Return if system reset has been issued
+  if (sys.abort)
+    return;
+
+  // Move in a line to the target
+  mc_line(target, feed_rate, invert_feed_rate, line_number);
+
+  // TODO: If the probe is already activated, we should look in
+  // the oppostie direction that is specified.
+
+  if (sensor == MAG_SENSOR)
+    probe.carousel_probe_state = PROBE_ACTIVE;
+
+  // Tell the system we are probing
+  probe.isprobing = 1;
+
+  sys.state = STATE_PROBING;
+  
+  if (!probe_loop())
+    return;
+
+  uint8_t probe_fail;
+  if (sensor == MAG_SENSOR) {
+    probe_fail = (probe.carousel_probe_state == PROBE_ACTIVE);
+    if (probe_fail)
+      memcpy(sys.probe_position, sys.position, sizeof(float) * N_AXIS);
+  }
+
+  protocol_execute_runtime();
+
+  if (sys.abort)
+    return;
+  
+  // Prep the new target based on the positon that the probe triggered
+  uint8_t idx;
```

Comment:
Sidenote: We really need to turn on C99 so we can stop requiring external iterators and widely scoped variables... I'll make a ticket for this.